aten/src/ATen/core/List.h

```diff
@@ -78,6 +78,10 @@ class ListElementReference final {
   // assigning another ref to this assigns the underlying value
   ListElementReference& operator=(ListElementReference&& rhs) &&;
 
+  const IValue& get() const& {
+    return *iterator_;
+  }
```

Comment:
idk if this is kosher or not. dare I ping smessmer or zdevito haha

Reply:
Well, I agree this seems necessary.

There is a public conversion to T (aka Tensor) which seems to be what ordinarily is supposed to be used. It looks like there is a refcount bump here, as we are obligated to produce an owning Tensor per this type signature, which cannot be done without refcount bumps unless you move out of the IValue. So it seems to me that the preexisting iterator interface over List will end up refcount bumping all your Tensors even if you say `for (const Tensor& t : tensor_list)`. That seems bad! Not sure if it is fixable though, maybe an explicit `operator const T&() const` would help?

cc @swolchok

Reply:
Doesn't look so good https://godbolt.org/z/bqh4q35da

Reply:
Just clarifying, here: so, is it okay to keep this?

Reply:
> That seems bad!

Yes it does! I am not a fan of List and Dict for efficiency reasons.

You might try to *replace* `operator T() const` by `operator const T&() const`, but IIRC the reason you can't do that is that `IValue::to()` usually cannot give you a reference (per ivalue_to.h).

Here's a good idea: you could instead specialize `ListElementReference` such that when `std::is_reference<c10::detail::ivalue_to_const_ref_overload_return<T>>::value` is `true`, it defines `operator const T&()`, and otherwise it defines `operator T()`. This would allow you to eliminate `get()` and also have the fantastic side effect of transparently fixing `for (const Tensor& t : tensor_list)` everywhere for free. Please pester me as much as necessary for help with that, because it sounds really appealing at first blush. (disclaimer: I've had to go cold turkey on caffeine very recently and might be talking nonsense!)

That said, I don't view `get()` as particularly *bad*.

Reply:
Yeah, let's keep it as is here.

Reply:
I think that's a good idea. But, there's another problem: `List<optional<Tensor>>`. `ivalue_to_const_ref_overload_return` is false for `optional<Tensor>`. Meaning that there would be a ref-count bump there. With `get()` we are able to avoid that by directly wrapping that into an `OptionalTensorRef`.